### PR TITLE
Optimize ChunkedBitSet dense relations

### PR DESCRIPTION
**Pull request by martingms**
`ChunkedBitSet::subtract(&HybridBitSet)` showed up in a callgrind profile grabbed of the `helloworld` benchmark while playing around with [rustc-perf](https://github.com/rust-lang/rustc-perf), and the FIXMEs left behind there by @nnethercote pointed to a way to optimize it.

The resulting improvements on the `rustc-perf` benchmarks are [meager](https://user-images.githubusercontent.com/458783/156841723-d89230d8-dfce-4fb5-a66d-c23737505bd5.png) to say the least, but perhaps the changes are of interest anyway.

I'm unacquainted with the compiler code and the tooling in general so I apologize in advance for any missteps!

PS: Sorry for the double PR, the first one was supposed to be a draft!

r? @nnethercote

### CODE REVIEW

**Comment by nnethercote** [comment]
Let's do a perf run on CI before I review this.

@bors try @rust-timer queue

File: compiler/rustc_index/src/bit_set.rs [hunk 1]
```diff
@@ -651,24 +651,174 @@ impl<T: Idx> BitRelations<ChunkedBitSet<T>> for ChunkedBitSet<T> {
...
         unimplemented!("implement if/when necessary");
     }
 }
 
+impl<T: Idx> BitRelations<BitSet<T>> for ChunkedBitSet<T> {
+    fn union(&mut self, other: &BitSet<T>) -> bool {
+        assert_eq!(self.domain_size, other.domain_size());
+
+        let mut changed = false;
+        for (chunk, other_words) in self.chunks.iter_mut().zip(other.words().chunks(CHUNK_WORDS)) {
```
Comment (nnethercote) [inline compiler/rustc_index/src/bit_set.rs:681]: Where is this `chunks()` method defined?

**Comment by nnethercote** [comment]
Thank you for the pull request! Well done on navigating the contribution process and completing it, everything looks fine in terms of how you've gone about it. I particularly like that you split it up into three logical commits. Hopefully we can look forward to more contributions from you in the future 😄 

I admit I am torn on this one. On one hand, the code looks good, much how I would expect. On the other hand, it's a bunch of extra lines of code and corresponding complexity that doesn't provide any measurable performance benefit. I have written plenty of PRs that I thought would help performance and then didn't, and ended up closing them, and my sense of engineering trade-offs is pushing me to do the same thing here. Sorry about that. One thing you could do is file another PR that adds comments to `union` and `subtract`, pointing at this PR, so this code is findable in the future should we identify a situation where faster versions of these operations would make a measurable difference. Does that make sense?

File: compiler/rustc_index/src/bit_set.rs (hunk 1 above)
Comment (martingms) [reply compiler/rustc_index/src/bit_set.rs:681]: https://doc.rust-lang.org/std/primitive.slice.html#method.chunks :+1:

**Comment by martingms** [comment]
> Sorry about that.

Don't be! I just needed something to work on at a hack night to improve my rust skills, no reason to merge unless it adds value.

> One thing you could do is file another PR that adds comments to `union` and `subtract`, pointing at this PR, so this code is findable in the future should we identify a situation where faster versions of these operations would make a measurable difference. Does that make sense?

Sure I can do that :)

**Pull request closed without merging**